src/lib.ts

```diff
@@ -683,3 +687,80 @@ export function printEventLog(): void {
     );
   }
 }
+
+const rateCache = new Map<string, Map<Monster, number>>();
+
+export function barfEncounterRate(options: {
+  snapper?: boolean;
+  snapperPhylum?: Phylum;
+  olfact?: Monster;
+  longCon?: Monster;
+  motif?: Monster;
+  turtle?: Monster;
+  monkeyPoint?: Monster;
+  humanity?: boolean;
+}): Map<Monster, number> {
+  const olfact = options.olfact ?? get("olfactedMonster");
+  const longCon = options.longCon ?? get("longConMonster");
+  const motif = options.motif ?? get("motifMonster");
+  const turtle = options.turtle ?? get("_gallapagosMonster");
+  const monkeyPoint = options.monkeyPoint ?? get("monkeyPointMonster");
+  const snapper = options.snapper ?? myFamiliar() === $familiar`Red-Nosed Snapper`;
+  const snapperPhylum = options.snapperPhylum ?? Snapper.getTrackedPhylum();
+  const humanity = options.humanity ?? have($effect`Ew, The Humanity`);
+
+  const zoneMonsters = $monsters`garbage tourist, angry tourist, horrible tourist family`;
+  const encounterQueue = zoneMonsters.filter((m) =>
+    $location`Barf Mountain`.combatQueue.includes(`${m}`),
+  );
+
+  const cacheKey = [
+    olfact,
+    longCon,
+    motif,
+    turtle,
+    monkeyPoint,
+    snapper,
+    snapperPhylum,
+    humanity,
+    ...encounterQueue,
+  ]
+    .map((v) => `${v}`)
+    .join(":");
+
+  const cachedValue = rateCache.get(cacheKey);
+  if (cachedValue) {
+    return cachedValue;
+  }
+
+  const copies = (target: Monster | null, n: number): Monster[] =>
+    n === 0 ? [] : [...zoneMonsters.filter((m) => m === target), ...copies(target, n - 1)];
```

Comment:
I feel like a `.fill` solution will be more performant and legible than the recursive one